tests/test_modules/test_contact.py

```diff
@@ -0,0 +1,274 @@
+
+
+class TestContact:
+    @pytest.fixture(scope="function")
+    def load_contact(self, request):
+        # contact between 15th and 23rd rod
+        return _Contact(15, 23)
+
+    @pytest.mark.parametrize("illegal_contact", [int, list])
+    def test_using_with_illegal_contact_throws_assertion_error(
+        self, load_contact, illegal_contact
+    ):
+        with pytest.raises(AssertionError) as excinfo:
+            load_contact.using(illegal_contact)
+        assert "not a valid contact class" in str(excinfo.value)
+
+    from elastica.contact_forces import NoContact
+
+    # TODO Add other legal contact later
+    @pytest.mark.parametrize("legal_contact", [NoContact])
+    def test_using_with_legal_contact(self, load_contact, legal_contact):
+        contact = load_contact
+        contact.using(legal_contact, 3, 4.0, "5", k=1, l_var="2", j=3.0)
+
+        assert contact._contact_cls == legal_contact
+        assert contact._args == (3, 4.0, "5")
+        assert contact._kwargs == {"k": 1, "l_var": "2", "j": 3.0}
+
+    def test_id(self, load_contact):
+        contact = load_contact
+        # This is purely for coverage purposes, no actual test
+        # since its a simple return
+        assert contact.id() == (15, 23)
+
+    def test_call_without_setting_contact_throws_runtime_error(self, load_contact):
+        contact = load_contact
+
+        with pytest.raises(RuntimeError) as excinfo:
+            contact()
+        assert "No contacts provided" in str(excinfo.value)
+
+    def test_call_improper_args_throws(self, load_contact):
+        # Example of bad initiailization function
+        # This needs at least four args which the user might
+        # forget to pass later on
+        def mock_init(self, *args, **kwargs):
+            self.nu = args[3]  # Need at least four args
+            self.k = kwargs.get("k")
+
+        # in place class
+        MockContact = type(
+            "MockContact", (self.NoContact, object), {"__init__": mock_init}
+        )
+
+        # The user thinks 4.0 goes to nu, but we don't accept it because of error in
+        # construction og a Contact class
+        contact = load_contact
+        contact.using(MockContact, 4.0, k=1, l_var="2", j=3.0)
+
+        # Actual test is here, this should not throw
+        with pytest.raises(TypeError) as excinfo:
+            _ = contact()
+        assert "Unable to construct" in str(excinfo.value)
+
+
+class TestContactMixin:
+    from elastica.modules import BaseSystemCollection
+
+    class SystemCollectionWithContactMixedin(BaseSystemCollection, Contact):
+        pass
+
+    from elastica.rod import RodBase
+    from elastica.rigidbody import RigidBodyBase
+    from elastica.surface import SurfaceBase
+
+    class MockRod(RodBase):
+        def __init__(self, *args, **kwargs):
+            self.n_elems = 3  # arbitrary number
+
+        # Contacts assume that this promise is met
+        def __len__(self):
+            return 2  # a random number
```

Comment:
Why do we need to set `__len__` ? Can you elaborate

Reply:
I got it from test_connections.py. It did not seem to do anything there too, so I assumed it is needed for some backend thing. Should I remove it from both and see what happens?

Reply:
yes can you test